File: app/src/main/java/de/garten/training/depotflow/data/api/FakeDepotApiClient.java
package de.garten.training.depotflow.data.api;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import de.garten.training.depotflow.core.ResultCallback;
import de.garten.training.depotflow.data.api.dto.ChecklistItemDto;
import de.garten.training.depotflow.data.api.dto.OutboxDto;
import de.garten.training.depotflow.data.api.dto.StopDto;
import de.garten.training.depotflow.data.api.dto.SyncResponseDto;
import de.garten.training.depotflow.data.api.dto.WorkOrderDto;

public class FakeDepotApiClient implements DepotApiClient {

    @Override
    public void loadWorkOrders(String depotId, String changedSince, ResultCallback<List<WorkOrderDto>> callback) {
        callback.onSuccess(Arrays.asList(
                order("wo-1001", "DF-2026-1001", "Torprüfung und Scanner-Tausch", "Norddepot Cloppenburg", "accepted", 8, "m.schneider"),
                order("wo-1002", "DF-2026-1002", "Ersatzteile an Außenlager liefern", "Servicepunkt Bremen", "in_progress", 5, "l.koenig"),
                order("wo-1003", "DF-2026-1003", "Temperatursensor im Außenlager prüfen", "Kühlbereich Oldenburg", "new", 9, "a.weber")
        ));
    }

    @Override
    public void loadWorkOrder(String serverId, ResultCallback<WorkOrderDto> callback) {
        callback.onSuccess(order(serverId, "DF-2026-DEMO", "Remote-Detail nachladen", "Demo-Depot", "accepted", 3, "demo.user"));
    }

    @Override
    public void completeWorkOrder(String serverId, OutboxDto payload, ResultCallback<WorkOrderDto> callback) {
        WorkOrderDto completed = order(serverId, "DF-2026-COMPLETE", "Lokal erledigter Auftrag", "Demo-Depot", "done", 1, "demo.user");
        completed.updatedAt = "2026-05-27T14:40:00Z";
        callback.onSuccess(completed);
    }

    @Override
    public void pushOutbox(List<OutboxDto> payload, ResultCallback<SyncResponseDto> callback) {
        SyncResponseDto response = new SyncResponseDto();
        response.success = true;
        response.acceptedAggregateIds = new ArrayList<>();
        response.rejectedAggregateIds = new ArrayList<>();
        response.serverTime = "2026-05-27T14:45:00Z";
        if (payload != null) {
            for (OutboxDto entry : payload) {
                response.acceptedAggregateIds.add(entry.aggregateId);
            }
        }
        callback.onSuccess(response);
    }

    private WorkOrderDto order(String id, String externalNumber, String title, String customer, String status, int priority, String user) {
        WorkOrderDto dto = new WorkOrderDto(id, externalNumber, title, customer, status, priority);
        dto.dueAt = "2026-05-28T10:00:00Z";
        dto.updatedAt = "2026-05-27T14:00:00Z";
        dto.assignedUser = user;
        dto.stops = Arrays.asList(
                stop(id + "-st-1", 1, "pickup", "Zentrallager", "Am Hafen 4", "done"),
                stop(id + "-st-2", 2, "delivery", customer, "Industriestraße 12", "open")
        );
        dto.checklist = Arrays.asList(
                checklist(id + "-ci-1", "Übergabe dokumentieren", false, true),
                checklist(id + "-ci-2", "Foto im Auftrag ablegen", false, false)
        );
        return dto;
    }

    private StopDto stop(String id, int sequence, String type, String name, String address, String status) {
        StopDto dto = new StopDto();
        dto.id = id;
        dto.sequence = sequence;
        dto.type = type;
        dto.name = name;
        dto.address = address;
        dto.latitude = 52.84 + sequence;
        dto.longitude = 8.04 + sequence;
        dto.status = status;
        dto.arrivalWindowFrom = "2026-05-28T08:00:00Z";
        dto.arrivalWindowTo = "2026-05-28T10:00:00Z";
        return dto;
    }

    private ChecklistItemDto checklist(String id, String label, boolean checked, boolean mandatory) {
        ChecklistItemDto dto = new ChecklistItemDto();
        dto.id = id;
        dto.label = label;
        dto.checked = checked;
        dto.mandatory = mandatory;
        dto.note = null;
        return dto;
    }
}

File: app/src/main/java/de/garten/training/depotflow/data/api/FakeDepotApiClient.kt
package de.garten.training.depotflow.data.api

import de.garten.training.depotflow.core.ResultCallback
import de.garten.training.depotflow.data.api.dto.ChecklistItemDto
import de.garten.training.depotflow.data.api.dto.OutboxDto
import de.garten.training.depotflow.data.api.dto.StopDto
import de.garten.training.depotflow.data.api.dto.SyncResponseDto
import de.garten.training.depotflow.data.api.dto.WorkOrderDto

class FakeDepotApiClient : DepotApiClient {

    override fun loadWorkOrders(
        depotId: String,
        changedSince: String?,
        callback: ResultCallback<List<WorkOrderDto>>,
    ) {
        callback.onSuccess(
            listOf(
                order(
                    id = "wo-1001",
                    externalNumber = "DF-2026-1001",
                    title = "Torprüfung und Scanner-Tausch",
                    customer = "Norddepot Cloppenburg",
                    status = "accepted",
                    priority = 8,
                    user = "m.schneider",
                ),
                order(
                    id = "wo-1002",
                    externalNumber = "DF-2026-1002",
                    title = "Ersatzteile an Außenlager liefern",
                    customer = "Servicepunkt Bremen",
                    status = "in_progress",
                    priority = 5,
                    user = "l.koenig",
                ),
                order(
                    id = "wo-1003",
                    externalNumber = "DF-2026-1003",
                    title = "Temperatursensor im Außenlager prüfen",
                    customer = "Kühlbereich Oldenburg",
                    status = "new",
                    priority = 9,
                    user = "a.weber",
                ),
            ),
        )
    }

    override fun loadWorkOrder(serverId: String, callback: ResultCallback<WorkOrderDto>) {
        callback.onSuccess(
            order(
                id = serverId,
                externalNumber = "DF-2026-DEMO",
                title = "Remote-Detail nachladen",
                customer = "Demo-Depot",
                status = "accepted",
                priority = 3,
                user = "demo.user",
            ),
        )
    }

    override fun completeWorkOrder(
        serverId: String,
        payload: OutboxDto,
        callback: ResultCallback<WorkOrderDto>,
    ) {
        callback.onSuccess(
            order(
                id = serverId,
                externalNumber = "DF-2026-COMPLETE",
                title = "Lokal erledigter Auftrag",
                customer = "Demo-Depot",
                status = "done",
                priority = 1,
                user = "demo.user",
            ).apply {
                updatedAt = "2026-05-27T14:40:00Z"
            },
        )
    }

    override fun pushOutbox(
        payload: List<OutboxDto>,
        callback: ResultCallback<SyncResponseDto>,
    ) {
        callback.onSuccess(
            SyncResponseDto().apply {
                success = true
                acceptedAggregateIds = payload.mapNotNull { it.aggregateId }.toMutableList()
                rejectedAggregateIds = mutableListOf()
                serverTime = "2026-05-27T14:45:00Z"
            },
        )
    }

    private fun order(
        id: String,
        externalNumber: String,
        title: String,
        customer: String,
        status: String,
        priority: Int,
        user: String,
    ) = WorkOrderDto(
        id = id,
        externalNumber = externalNumber,
        title = title,
        customerName = customer,
        status = status,
        priority = priority,
    ).apply {
        dueAt = DEFAULT_DUE_AT
        updatedAt = DEFAULT_UPDATED_AT
        assignedUser = user
        stops = listOf(
            stop(
                id = "$id-st-1",
                sequence = 1,
                type = "pickup",
                name = "Zentrallager",
                address = "Am Hafen 4",
                status = "done",
            ),
            stop(
                id = "$id-st-2",
                sequence = 2,
                type = "delivery",
                name = customer,
                address = "Industriestraße 12",
                status = "open",
            ),
        )
        checklist = listOf(
            checklist(id = "$id-ci-1", label = "Übergabe dokumentieren", checked = false, mandatory = true),
            checklist(id = "$id-ci-2", label = "Foto im Auftrag ablegen", checked = false, mandatory = false),
        )
    }

    private fun stop(
        id: String,
        sequence: Int,
        type: String,
        name: String,
        address: String,
        status: String,
    ) = StopDto().apply {
        this.id = id
        this.sequence = sequence
        this.type = type
        this.name = name
        this.address = address
        latitude = BASE_LATITUDE + sequence
        longitude = BASE_LONGITUDE + sequence
        this.status = status
        arrivalWindowFrom = DEFAULT_ARRIVAL_FROM
        arrivalWindowTo = DEFAULT_ARRIVAL_TO
    }

    private fun checklist(
        id: String,
        label: String,
        checked: Boolean,
        mandatory: Boolean,
    ) = ChecklistItemDto().apply {
        this.id = id
        this.label = label
        this.checked = checked
        this.mandatory = mandatory
    }

    private companion object {
        const val DEFAULT_DUE_AT = "2026-05-28T10:00:00Z"
        const val DEFAULT_UPDATED_AT = "2026-05-27T14:00:00Z"
        const val DEFAULT_ARRIVAL_FROM = "2026-05-28T08:00:00Z"
        const val DEFAULT_ARRIVAL_TO = "2026-05-28T10:00:00Z"
        const val BASE_LATITUDE = 52.84
        const val BASE_LONGITUDE = 8.04
    }
}
